internal/engine/experiment/tlsmiddlebox/tracing.go

```diff
@@ -0,0 +1,165 @@
+package tlsmiddlebox
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/engine/experiment/tlsmiddlebox/internal"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+// MeasureTLS outputs a TraceEvent with the iterative trace for the passSNI and the targetSNI
+func (m *Measurer) MeasureTLS(ctx context.Context, addr string, targetSNI string, tlsEvents chan<- *CompleteTrace) {
+	out := &CompleteTrace{}
+	passSNI := m.config.snipass()
+	passTrace := m.IterativeTrace(ctx, addr, passSNI)
+	out.PassTrace = passTrace
+	targetTrace := m.IterativeTrace(ctx, addr, targetSNI)
+	out.TargetTrace = targetTrace
+	select {
+	case tlsEvents <- out:
+	default:
+		return
+	}
+}
+
+// IterativeTrace calls the iterativeTrace and populates the TraceEvent with iteration results
+func (m *Measurer) IterativeTrace(ctx context.Context, addr string, sni string) (trace *TraceEvent) {
+	iterations := m.config.iterations()
+	out := make(chan *IterEvent, iterations)
+	trace = &TraceEvent{
+		Address:    addr,
+		SNI:        sni,
+		Iterations: []*IterEvent{},
+	}
+	m.iterativeTrace(ctx, addr, sni, iterations, out)
+	iterEvents := extractEvents(out) // align the iteration results before modeling them
+	trace.AddIterations(iterEvents)
+	return trace
+}
+
+func (m *Measurer) iterativeTrace(ctx context.Context, addr string, sni string,
+	iterations int, out chan<- *IterEvent) {
+	ticker := time.NewTicker(m.config.delay())
+	wg := new(sync.WaitGroup)
+	for i := 1; i <= iterations; i++ {
+		wg.Add(1)
+		go iterAsync(ctx, addr, sni, i, out, wg)
+		<-ticker.C
+	}
+	wg.Wait()
+}
+
+// Single Iteration for network tracing
+func iterAsync(ctx context.Context, addr string, sni string,
+	ttl int, out chan<- *IterEvent, wg *sync.WaitGroup) {
+	defer wg.Done()
+	select {
+	case out <- HandshakeWithTTL(ctx, addr, sni, ttl):
+	default:
+		return
+	}
+}
+
+// This handles the conn and calls the handshake function after setting the TTL value
+func HandshakeWithTTL(ctx context.Context, addr string, sni string, ttl int) (out *IterEvent) {
+	logger := model.ValidLoggerOrDefault(nil)
+	out = &IterEvent{
+		TTL:     ttl,
+		Failure: nil,
+	}
+	// we use the net.Dialer instead of netxlite.Dialer
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", addr)
+	if err != nil {
+		errStr := err.Error()
+		out.Failure = &errStr
+		return
+	}
+	defer conn.Close()
+	err = internal.SetConnTTL(conn, ttl)
+	if err != nil {
+		errStr := err.Error()
+		out.Failure = &errStr
+		return
+	}
+	performHandshake(ctx, conn, sni, logger, out)
```

Comment:
Maybe since you are dropping everything from the measurement after a successful handshake or when you get a reset, you can detect this early and release the wait group to avoid waiting on the successive traces.

Reply:
The delay is pretty short though, so it's perhaps not really a needed optimisation.